xds/src/test/java/io/grpc/xds/sds/SdsClientUdsTest.java

```diff
@@ -0,0 +1,212 @@
+
+  private TestSdsServer.SecretGetter secretGetter;
+  private TestSdsServer server;
+  private SdsClient sdsClient;
+  private Node node;
+  private SdsSecretConfig sdsSecretConfig;
+
+  private static ConfigSource buildConfigSource(String targetUri) {
+    return ConfigSource.newBuilder()
+        .setApiConfigSource(
+            ApiConfigSource.newBuilder()
+                .setApiType(ApiConfigSource.ApiType.GRPC)
+                .addGrpcServices(
+                    GrpcService.newBuilder()
+                        .setGoogleGrpc(
+                            GrpcService.GoogleGrpc.newBuilder().setTargetUri(targetUri).build())
+                        .build())
+                .build())
+        .build();
+  }
+
+  private static String getResourcesFileContent(String resFile) throws IOException {
+    String tempFile = TestUtils.loadCert(resFile).getAbsolutePath();
+    return new String(Files.readAllBytes(Paths.get(tempFile)), StandardCharsets.UTF_8);
+  }
+
+  @Before
+  public void setUp() throws IOException {
+    secretGetter = mock(TestSdsServer.SecretGetter.class);
+    server = new TestSdsServer(secretGetter);
+    server.startServer(SDSCLIENT_TEST_SOCKET, true);
+    ConfigSource configSource = buildConfigSource("unix:" + SDSCLIENT_TEST_SOCKET);
+    sdsSecretConfig =
+        SdsSecretConfig.newBuilder().setSdsConfig(configSource).setName("name1").build();
+    node = Node.newBuilder().setId("sds-client-temp-test2").build();
+    sdsClient = new SdsClient(sdsSecretConfig, node);
+    sdsClient.start(MoreExecutors.directExecutor());
+  }
+
+  @After
+  public void teardown() {
+    sdsClient.shutdown();
+    server.shutdown();
+  }
+
+  @Test
+  public void testSecretWatcher_tlsCertificate() throws IOException {
+    SdsClient.SecretWatcher mockWatcher = mock(SdsClient.SecretWatcher.class);
+
+    when(secretGetter.getFor("name1"))
+        .thenReturn(getOneTlsCertSecret("name1", SERVER_0_KEY_FILE, SERVER_0_PEM_FILE));
+
+    sdsClient.watchSecret(mockWatcher);
+    discoveryRequestVerification(server.lastGoodRequest, "[name1]", "", "");
+    discoveryRequestVerification(
+        server.lastRequestOnlyForAck,
+        "[name1]",
+        server.lastResponse.getVersionInfo(),
+        server.lastResponse.getNonce());
+    secretWatcherVerification(mockWatcher, "name1", SERVER_0_KEY_FILE, SERVER_0_PEM_FILE);
+
+    reset(mockWatcher);
+    when(secretGetter.getFor("name1"))
+        .thenReturn(getOneTlsCertSecret("name1", SERVER_1_KEY_FILE, SERVER_1_PEM_FILE));
+    server.generateAsyncResponse("name1");
+    secretWatcherVerification(mockWatcher, "name1", SERVER_1_KEY_FILE, SERVER_1_PEM_FILE);
+
+    reset(mockWatcher);
+    sdsClient.cancelSecretWatch(mockWatcher);
+    server.generateAsyncResponse("name1");
+    verify(mockWatcher, never()).onSecretChanged(ArgumentMatchers.any(Secret.class));
+  }
+
+  private void discoveryRequestVerification(
```

Comment:
for method name it should be verb. verifyDiscoveryRequest or assertDisvoeryRequest since it includes assertion.